akka-actor-typed/src/main/scala/akka/actor/typed/receptionist/Receptionist.scala

```diff
@@ -110,24 +129,124 @@ object Receptionist extends ExtensionId[Receptionist] {
   @InternalApi
   private[typed] abstract class InternalCommand extends AllCommands
 
+  /**
+   * Internal API
+   */
+  @InternalApi
+  private[akka] object MessageImpls {
+    // some trixery here to provide a nice _and_ safe API in the face
+    // of type erasure, more type safe factory methods for each message
+    // is the user API below while still hiding the type parameter so that
+    // users don't incorrecly match against it
+
+    final case class Register[T] private[akka] (
+      key:             ServiceKey[T],
+      serviceInstance: ActorRef[T],
+      replyTo:         Option[ActorRef[Receptionist.Registered]]) extends Command
+
+    final case class Registered[T] private[akka] (key: ServiceKey[T], _serviceInstance: ActorRef[T]) extends Receptionist.Registered {
+      def isForKey(key: ServiceKey[_]): Boolean = key == this.key
+      def serviceInstance[M](key: ServiceKey[M]): ActorRef[M] = {
+        if (key != this.key) throw new IllegalArgumentException(s"Wrong key [$key] used, must use listing key [${this.key}]")
```

Comment:
So we want people to call `isForKey` first if they want to prevent running into the exception?

Reply:
Yep, on the Java side, on the Scala side I have updated all sampels to just use `key.Registered(ref)`